File: packages/engine/src/cli/host.ts
import { resolve } from 'node:path'
import type { HookResult } from './event'

/**
 * Hook hosts (r1-fixes 6.3–6.6, D34; agents-parity 2.1, D5): ONE set of
 * handlers serves Claude Code, Cursor and Codex, and this module is the whole
 * difference between them.
 *
 * Cursor runs the same shim files — natively from .cursor/hooks.json, or as
 * "third-party" hooks read out of .claude/settings.json — but it speaks its own
 * dialect on both ends of the pipe, and a handler written for Claude Code's
 * dialect fails there SILENTLY rather than loudly:
 *
 * - IN: Cursor hands its own payload even to a Claude-format hook. Tools are
 *   named `Shell` and `Write` (never `Bash` / `Edit`), a failure carries
 *   `error_message`, and the stop hook's loop guard is `loop_count`. Round 1's
 *   Cursor cells show the cost: 175 file_touched (Write happens to match) and
 *   0 command_run (Shell never matched Bash).
 * - OUT: Cursor ignores plain stdout and does nothing with exit 2 on stop. It
 *   reads `additional_context` and `followup_message` from a JSON object.
 *   Cursor can translate Claude's `hookSpecificOutput` too, but only behind a
 *   compatibility flag sofar cannot see, so the native keys are emitted.
 *
 * Everything here was read from cursor-agent 2026.09.10-fd3934a, and SPEC
 * pins it (§Cursor host). The handlers stay Claude-shaped and byte-identical for Claude Code:
 * `forHost` converts a Cursor payload on the way in and the result on the way
 * out, and a Claude Code invocation passes straight through.
 *
 * Codex (read from codex-cli 0.154.0 and its docs, SPEC §Codex host) already
 * sends Claude Code's field names, so its payload needs no conversion. What
 * it lacks is any field naming the host, so its shims declare it
 * (`--host codex`, D5). Three differences remain (D6): `apply_patch` edits
 * (parsed by `patchedFiles`), a PostToolUse that also fires after a failing
 * command, and context carried as `hookSpecificOutput` JSON so the digest can
 * be exempted from Codex's 2,500-token spill.
 */

type Obj = Record<string, unknown>

/** The hook subcommand names (event.ts SUBCOMMANDS) — the unit a conversion is chosen by. */
export type HookName = 'session-start' | 'post-tool' | 'post-tool-failure' | 'user-prompt' | 'stop' | 'session-end'

/** Which agent fired a hook — recorded on session registration and diagnostics rows. */
export interface HookHost {
  tool: 'claude-code' | 'cursor' | 'codex'
  /** The host's own version string, when its payload names one. */
  version?: string
}

export const CLAUDE_CODE_HOST: HookHost = { tool: 'claude-code' }
export const CODEX_HOST: HookHost = { tool: 'codex' }

/**
 * Hosts whose payload names no host, so their shims must, as
 * `sofar event <hook> --host <id>` (D5). A host that can be told from stdin
 * never needs the flag.
 */
export const DECLARED_HOSTS = ['codex'] as const
export type DeclaredHost = (typeof DECLARED_HOSTS)[number]

export function isDeclaredHost(value: string): value is DeclaredHost {
  return (DECLARED_HOSTS as readonly string[]).includes(value)
}

/**
 * Does this host fire PostToolUse only for a call that succeeded? Claude Code
 * and Cursor do, and send failures to PostToolUseFailure. Codex has no failure
 * event and fires PostToolUse after a Bash command that exits non-zero as well
 * (docs), with no verified exit status on the payload. So a Codex firing says
 * nothing about the outcome, and the event records none: absent `ok` means
 * unknown, never success.
 */
export function postToolProvesSuccess(host: HookHost): boolean {
  return host.tool !== 'codex'
}

/**
 * Cursor's per-carrier cap on injected context, measured after trimming. Above
 * it Cursor drops the WHOLE carrier rather than truncating it, so a notice one
 * character too long delivers nothing at all.
 */
export const CURSOR_CONTEXT_MAX = 10_000

/** Cursor's hook tool names → the Claude Code names the handlers classify. */
const CURSOR_TOOL_NAMES: Readonly<Record<string, string>> = { Shell: 'Bash' }

function isObj(v: unknown): v is Obj {
  return typeof v === 'object' && v !== null && !Array.isArray(v)
}

/**
 * Which host fired this hook. Read from STDIN only — `cursor_version` is on
 * every Cursor payload — and never from the environment: a Claude Code session
 * started inside Cursor's integrated terminal inherits Cursor's variables and
 * must still be served as Claude Code.
 */
export function hookHost(hook: Obj): HookHost {
  const version = hook.cursor_version
  if (typeof version !== 'string') return CLAUDE_CODE_HOST
  return version.length > 0 ? { tool: 'cursor', version } : { tool: 'cursor' }
}

/**
 * A Cursor payload in the field names the handlers read. Every original field
 * is kept (hookHost still has to see `cursor_version`); an alias is added only
 * where the Claude name is absent, so nothing Cursor sent is overwritten.
 */
export function fromCursor(hook: Obj): Obj {
  const out: Obj = { ...hook }
  if (typeof hook.session_id !== 'string' && typeof hook.conversation_id === 'string') {
    out.session_id = hook.conversation_id
  }
  if (typeof hook.tool_name === 'string') {
    const mapped = CURSOR_TOOL_NAMES[hook.tool_name]
    if (mapped !== undefined) out.tool_name = mapped
  }
  if (typeof hook.error_message === 'string' && hook.error === undefined) out.error = hook.error_message
  if (typeof hook.tool_output === 'string' && hook.tool_response === undefined) {
    out.tool_response = { stdout: hook.tool_output }
  }
  // Cursor counts the follow-ups a stop hook has already caused; Claude Code
  // flags the same state as a boolean. Either way: we have held once already.
  if (typeof hook.loop_count === 'number' && hook.stop_hook_active === undefined) {
    out.stop_hook_active = hook.loop_count > 0
  }
  return out
}

/** The context a handler's stdout carries, whichever Claude Code form it took. */
function contextOf(name: HookName, stdout: string): string | null {
  const text = stdout.trim()
  if (text.length === 0) return null
  if (name !== 'post-tool') return text
  // PostToolUse speaks hookSpecificOutput JSON for Claude Code.
  try {
    const decoded: unknown = JSON.parse(text)
    const specific = isObj(decoded) ? decoded.hookSpecificOutput : undefined
    const context = isObj(specific) ? specific.additionalContext : undefined
    return typeof context === 'string' && context.trim().length > 0 ? context : null
  } catch {
    return null
  }
}

function json(value: Obj): string {
  return `${JSON.stringify(value)}\n`
}

/**
 * A handler's Claude Code result, as Cursor reads it. The Stop gate's exit 2
 * becomes exit 0 with `followup_message` — the only stop output Cursor acts on,
 * queued as the agent's next prompt. Context becomes `additional_context`.
 * Session-start is the one carrier Cursor does not cap, and the digest is
 * already held to 10,000 characters, so only the per-prompt and per-tool lines
 * are clipped — and clipped, never dropped, because Cursor's own overflow
 * handling is to discard the lot.
 */
export function toCursor(name: HookName, result: HookResult): HookResult {
  if (name === 'stop') {
    if (result.exitCode !== 2) return result
    const message = result.stderr.trim()
    return { exitCode: 0, stdout: message.length > 0 ? json({ followup_message: message }) : '', stderr: '' }
  }
  const context = contextOf(name, result.stdout)
  if (context === null) return { ...result, stdout: '' }
  const clipped =
    name === 'session-start' || context.length <= CURSOR_CONTEXT_MAX
      ? context
      : `${context.slice(0, CURSOR_CONTEXT_MAX - 1)}…`
  return { ...result, stdout: json({ additional_context: clipped }) }
}

/** The Codex event whose context carrier each context-bearing hook fills. */
const CODEX_CONTEXT_EVENTS: Readonly<Partial<Record<HookName, string>>> = {
  'session-start': 'SessionStart',
  'user-prompt': 'UserPromptSubmit',
}

/**
 * A handler's Claude Code result, as Codex reads it. Codex takes plain stdout
 * as context on SessionStart and UserPromptSubmit, but the docs say nothing on
 * whether plain text is exempt from its ~2,500-token spill, and a 10,000-char
 * digest sits right at that line. A spilled digest reaches the model as a
 * head-and-tail preview. So both carry `hookSpecificOutput.additionalContext`,
 * the one form `additionalContextLimit` (0 on sofar's SessionStart entry) is
 * documented to govern. Post-tool JSON, the Stop gate's exit 2 and every
 * empty result already fit Codex's schemas and pass through.
 */
export function toCodex(name: HookName, result: HookResult): HookResult {
  const event = CODEX_CONTEXT_EVENTS[name]
  if (event === undefined) return result
  const context = contextOf(name, result.stdout)
  if (context === null) return { ...result, stdout: '' }
  return { ...result, stdout: json({ hookSpecificOutput: { hookEventName: event, additionalContext: context } }) }
}

/**
 * The file headers of an `apply_patch` body, which Codex's edits carry whole
 * in `tool_input.command` with no `file_path`. The markers were read from codex
 * 0.154.0's binary; the grammar around them is unverified (SPEC §Codex host).
 */
const PATCH_HEADERS: ReadonlyArray<readonly [marker: string, op: string]> = [
  ['*** Add File: ', 'write'],
  ['*** Update File: ', 'edit'],
  ['*** Delete File: ', 'delete'],
]
const PATCH_MOVE = '*** Move to: '

/**
 * Every file one `apply_patch` touches, in patch order, resolved against the
 * session cwd the payload names (a patch path is relative to it, and Claude
 * Code's absolute `file_path` is the form the record already holds). A move
 * touches two paths: its source is gone (`delete`) and its destination is
 * written (`write`). Hunk lines never start with `***`, so a header cannot be
 * mistaken inside a hunk.
 */
export function patchedFiles(patch: string, cwd: string | null): Array<{ path: string; op: string }> {
  const at = (path: string): string => (cwd === null ? path : resolve(cwd, path))
  const files: Array<{ path: string; op: string }> = []
  for (const raw of patch.split('\n')) {
    const line = raw.endsWith('\r') ? raw.slice(0, -1) : raw
    if (line.startsWith(PATCH_MOVE)) {
      const to = line.slice(PATCH_MOVE.length).trim()
      const from = files[files.length - 1]
      if (to.length > 0 && from !== undefined && from.op === 'edit') {
        from.op = 'delete'
        files.push({ path: at(to), op: 'write' })
      }
      continue
    }
    for (const [marker, op] of PATCH_HEADERS) {
      if (!line.startsWith(marker)) continue
      const path = line.slice(marker.length).trim()
      if (path.length > 0) files.push({ path: at(path), op })
    }
  }
  return files
}

/**
 * Serve a hook handler to whichever host fired it. A host its shim declares
 * (Codex) is served as declared. Otherwise a payload that cannot be a Cursor
 * one (the substring check spares every Claude Code call a second parse on the
 * 100ms hook path) reaches the handler untouched.
 */
export function forHost(
  name: HookName,
  handler: (rootDir: string, input: string, host?: HookHost) => HookResult,
): (rootDir: string, input: string, declared?: DeclaredHost) => HookResult {
  return (rootDir, input, declared) => {
    if (declared === 'codex') return toCodex(name, handler(rootDir, input, CODEX_HOST))
    if (!input.includes('"cursor_version"')) return handler(rootDir, input)
    let hook: Obj
    try {
      const decoded: unknown = JSON.parse(input)
      if (!isObj(decoded)) return handler(rootDir, input)
      hook = decoded
    } catch {
      return handler(rootDir, input)
    }
    if (hookHost(hook).tool !== 'cursor') return handler(rootDir, input)
    return toCursor(name, handler(rootDir, JSON.stringify(fromCursor(hook))))
  }
}

File: crates/sofar-core/src/host.rs
//! Hook hosts (r1-fixes 6.3–6.6, D34) — the port of `cli/host.ts`: ONE set
//! of shims serves Claude Code and Cursor, and this module is the whole
//! difference. The handlers speak Claude Code's dialect; a Cursor payload
//! (any with a string `cursor_version`) is converted on the way in and the
//! result on the way out; a Claude Code invocation passes straight through.

use crate::cli::Hook;
use crate::fold_cli::CmdResult;
use crate::json::{self, Json, Object, stringify};
use crate::text::{js_trim, utf16_len};

/// Which agent fired a hook — recorded on session registration and diagnostics rows.
#[derive(Debug, Clone, PartialEq, Eq)]
pub struct HookHost {
    pub tool: &'static str,
    pub version: Option<String>,
}

pub const CLAUDE_CODE: &str = "claude-code";
pub const CURSOR: &str = "cursor";

/// Cursor's per-carrier cap on injected context (UTF-16 units, after trimming).
pub const CURSOR_CONTEXT_MAX: usize = 10_000;

/// `hookHost`: from STDIN only — `cursor_version` is on every Cursor payload.
#[must_use]
pub fn hook_host(hook: &Object) -> HookHost {
    match hook.get("cursor_version").and_then(Json::as_str) {
        None => HookHost {
            tool: CLAUDE_CODE,
            version: None,
        },
        Some(v) => HookHost {
            tool: CURSOR,
            version: if v.is_empty() {
                None
            } else {
                Some(v.to_owned())
            },
        },
    }
}

/// `fromCursor`: a Cursor payload in the field names the handlers read; every
/// original field kept, an alias added only where the Claude name is absent.
#[must_use]
pub fn from_cursor(hook: &Object) -> Object {
    let mut out = hook.clone();
    if hook.get("session_id").and_then(Json::as_str).is_none()
        && let Some(id) = hook.get("conversation_id").and_then(Json::as_str)
    {
        out.insert("session_id", Json::Str(id.to_owned()));
    }
    if hook.get("tool_name").and_then(Json::as_str) == Some("Shell") {
        out.insert("tool_name", Json::Str("Bash".to_owned()));
    }
    if let Some(message) = hook.get("error_message").and_then(Json::as_str)
        && hook.get("error").is_none()
    {
        out.insert("error", Json::Str(message.to_owned()));
    }
    if let Some(output) = hook.get("tool_output").and_then(Json::as_str)
        && hook.get("tool_response").is_none()
    {
        let mut response = Object::with_capacity(1);
        response.insert("stdout", Json::Str(output.to_owned()));
        out.insert("tool_response", Json::Obj(response));
    }
    if let Some(Json::Num(count)) = hook.get("loop_count")
        && hook.get("stop_hook_active").is_none()
    {
        out.insert("stop_hook_active", Json::Bool(*count > 0.0));
    }
    out
}

/// `contextOf`: the context a handler's stdout carries.
fn context_of(name: Hook, stdout: &str) -> Option<String> {
    let text = js_trim(stdout);
    if text.is_empty() {
        return None;
    }
    if name != Hook::PostTool {
        return Some(text.to_owned());
    }
    let Json::Obj(decoded) = json::parse(text).ok()? else {
        return None;
    };
    let context = decoded
        .get("hookSpecificOutput")
        .and_then(Json::as_obj)?
        .get("additionalContext")
        .and_then(Json::as_str)?;
    if js_trim(context).is_empty() {
        None
    } else {
        Some(context.to_owned())
    }
}

fn json_line(key: &str, value: &str) -> String {
    let mut o = Object::with_capacity(1);
    o.insert(key, Json::Str(value.to_owned()));
    format!("{}\n", stringify(&Json::Obj(o)))
}

/// `toCursor`: a handler's Claude Code result, as Cursor reads it.
#[must_use]
pub fn to_cursor(name: Hook, result: CmdResult) -> CmdResult {
    if name == Hook::Stop {
        if result.exit_code != 2 {
            return result;
        }
        let message = js_trim(&result.stderr);
        return CmdResult {
            exit_code: 0,
            stdout: if message.is_empty() {
                String::new()
            } else {
                json_line("followup_message", message)
            },
            stderr: String::new(),
        };
    }
    let Some(context) = context_of(name, &result.stdout) else {
        return CmdResult {
            stdout: String::new(),
            ..result
        };
    };
    let clipped = if name == Hook::SessionStart || utf16_len(&context) <= CURSOR_CONTEXT_MAX {
        context
    } else {
        let head: String = char::decode_utf16(context.encode_utf16().take(CURSOR_CONTEXT_MAX - 1))
            .map(|c| c.unwrap_or(char::REPLACEMENT_CHARACTER))
            .collect();
        format!("{head}…")
    };
    CmdResult {
        stdout: json_line("additional_context", &clipped),
        ..result
    }
}

/// `forHost`: serve a handler to whichever host fired it.
pub fn for_host(name: Hook, input: &str, handler: impl Fn(&str) -> CmdResult) -> CmdResult {
    if !input.contains("\"cursor_version\"") {
        return handler(input);
    }
    let Ok(Json::Obj(hook)) = json::parse(input) else {
        return handler(input);
    };
    if hook_host(&hook).tool != CURSOR {
        return handler(input);
    }
    to_cursor(name, handler(&stringify(&Json::Obj(from_cursor(&hook)))))
}

#[cfg(test)]
mod tests {
    use super::*;

    fn obj(text: &str) -> Object {
        match json::parse(text).unwrap() {
            Json::Obj(o) => o,
            _ => panic!("object"),
        }
    }

    #[test]
    fn cursor_payloads_are_aliased_never_overwritten() {
        let hook = obj(
            r#"{"cursor_version":"2026.09.10","conversation_id":"c1","tool_name":"Shell","error_message":"boom","tool_output":"out","loop_count":1}"#,
        );
        assert_eq!(hook_host(&hook).tool, CURSOR);
        let out = from_cursor(&hook);
        assert_eq!(out.get("session_id").and_then(Json::as_str), Some("c1"));
        assert_eq!(out.get("tool_name").and_then(Json::as_str), Some("Bash"));
        assert_eq!(out.get("error").and_then(Json::as_str), Some("boom"));
        assert_eq!(out.get("stop_hook_active"), Some(&Json::Bool(true)));
        assert_eq!(
            stringify(&Json::Obj(out)),
            r#"{"cursor_version":"2026.09.10","conversation_id":"c1","tool_name":"Bash","error_message":"boom","tool_output":"out","loop_count":1,"session_id":"c1","error":"boom","tool_response":{"stdout":"out"},"stop_hook_active":true}"#
        );
        assert_eq!(hook_host(&obj(r#"{"session_id":"s"}"#)).tool, CLAUDE_CODE);
    }

    #[test]
    fn results_speak_cursor() {
        let stop = CmdResult {
            exit_code: 2,
            stdout: String::new(),
            stderr: "write back\n".into(),
        };
        assert_eq!(
            to_cursor(Hook::Stop, stop).stdout,
            "{\"followup_message\":\"write back\"}\n"
        );
        let post = CmdResult { exit_code: 0, stdout: "{\"hookSpecificOutput\":{\"hookEventName\":\"PostToolUse\",\"additionalContext\":\"ctx\"}}\n".into(), stderr: String::new() };
        assert_eq!(
            to_cursor(Hook::PostTool, post).stdout,
            "{\"additional_context\":\"ctx\"}\n"
        );
        let empty = CmdResult {
            exit_code: 0,
            stdout: "  \n".into(),
            stderr: String::new(),
        };
        assert_eq!(to_cursor(Hook::UserPrompt, empty).stdout, "");
        let claude = for_host(Hook::Stop, "{\"session_id\":\"s\"}", |i| CmdResult {
            exit_code: 2,
            stdout: i.to_owned(),
            stderr: String::new(),
        });
        assert_eq!(claude.exit_code, 2);
    }
}
